chromadb/utils/data_loaders.py

```diff
@@ -22,3 +23,123 @@ def _load_image(self, uri: Optional[URI]) -> Optional[Image]:
     def __call__(self, uris: Sequence[Optional[URI]]) -> List[Optional[Image]]:
         with ThreadPoolExecutor(max_workers=self._max_workers) as executor:
             return list(executor.map(self._load_image, uris))
+
+# --------------------------- A lightweight util ---------------------------
+
+# TODO: Placing this here, since data loaders are a common use case for this function.
+#   But it should probably be moved to a more general-purpose module.
+#   See: https://github.com/chroma-core/chroma/issues/1606
+from functools import partial
+from concurrent.futures import ThreadPoolExecutor
+
+def vectorize(func, iterable=None, *, max_workers : int = 1):
+    """Like builtin map, but returns a list, 
+    and if iterable is None, returns a partial function that can directly be applied 
+    to an iterable.
+    This is useful, for instance, for making a data loader from any single-uri loader.
+    
+    Example:
+    >>> vectorize(lambda x: x**2, [1,2,3])
+    [1, 4, 9]
+    >>> vectorized_square = vectorize(lambda x: x**2)
+    >>> vectorized_square([1,2,3])
+    [1, 4, 9]
+    """
+    if iterable is None:
+        return partial(vectorize, func, max_workers=max_workers)
+    if max_workers == 1:
+        return list(map(func, iterable))
+    else:
+        with ThreadPoolExecutor(max_workers=max_workers) as executor:
+            return list(executor.map(func, iterable))
+        
+# --------------------------- Examples of loaders ---------------------------
+
+FileContents = Union[str, bytes]  # a type for the contents of a file
+
+def load_text(filepath: str) -> str:
```

Comment:
Perhaps you can tell me more of what you're thinking of. 

Here's what my thinking is (I'm assuming you are taking about making protocols for loader functions.

If we're talking file loaders, you could say that keys (a.k.a. paths) should be strings and contents can be strings or bytes (my `FileContents = Union[str, bytes]` type). But here's the thing: Maybe we want to use keys that are not strings, and perhaps we want to load other kinds of objects (lists, arrays, any python object...). 

So at the end of the (abstract) line, we just have `Callable[[Any], Any]`, or at a push, we can distinguish the key and the value roles by saying `Callable[[typing.KT], typing.VT]`. And if we have, for example, keys that are tuples, and values (what the loader returns) that are dicts, we just say `Callable[[tuple], dict]`. 

The protocol way would be (I'm guessing this is what you mean):

```python
from typing import TypeVar, Protocol, KT, VT, Union


class Loader(Protocol[KT, VT]):
    def meth(self, key: KT) -> VT:
        ...
        
# Examples:
PathType = str
FileContents = Union[bytes, str]
FileLoader = Loader[PathType, FileContents]
BytesFileLoader = Loader[PathType, bytes]
TextFileLoader = Loader[PathType, str]
```

I always swing from too much typing to not enough because I like the domain-driven naming and declarative nature of it, but it ends up being a lot of fluff whose advantage is not completely clear to me. 

Again -- what are your thoughts?

Reply:
I'll let you think about it and let me know if you want me to slap that in there too (and annotate the `loader` argument of `FileLoader` with it, I guess.).